src/block/request.rs

```diff
@@ -204,11 +256,142 @@ impl Request {
     }
 }
 
+/// Wraps a block device file for request execution.
+///
+/// # Example
+///
+/// ```rust
+/// # use std::fs::File;
+/// # use vm_memory::GuestAddress;
+/// # use vm_virtio::request::RequestExecutor;
+///
+/// let file = File::create("foo.txt").unwrap();
+/// let request_exec = RequestExecutor::new(file);
+/// ```
+pub struct RequestExecutor<D: DiskFile> {
+    /// The block device backing file.
+    inner: D,
+    /// The number of sectors of `inner`.
+    num_sectors: u64,
+}
+
+impl<D: DiskFile> Deref for RequestExecutor<D> {
+    type Target = D;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<D: DiskFile> DerefMut for RequestExecutor<D> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<D: DiskFile> RequestExecutor<D> {
+    /// Creates a new `RequestExecutor` based on `inner` object.
+    pub fn new(mut inner: D) -> Self {
+        let disk_size = inner.seek(SeekFrom::End(0)).unwrap() as u64;
+        // This check makes sense only if VIRTIO_BLK_F_BLK_SIZE feature is
+        // unsupported, which might be okay to assume for now.
+        // If we add support for this feature as well, this will have to be
+        // restricted to -> `if disk_size % blk_size != 0`.
+        // Also, do this check and num sectors computation make sense here?
+        // Should `new()` receive an already computed num sectors?
+        if disk_size % SECTOR_SIZE != 0 {
+            warn!(
+                "Disk size {} is not a multiple of sector size {}; \
+                 the remainder will not be visible to the guest.",
+                disk_size, SECTOR_SIZE
+            );
+        }
+
+        Self {
+            inner,
+            num_sectors: disk_size >> SECTOR_SHIFT,
+        }
+    }
+
+    fn num_sectors(&self) -> u64 {
+        self.num_sectors
+    }
+
+    fn len_to_num_sectors(len: u32) -> u64 {
+        let mut num_sectors = len as u64 / SECTOR_SIZE;
+        if len as u64 % SECTOR_SIZE != 0 {
+            num_sectors += 1;
+        }
+        num_sectors
+    }
+
+    fn check_access(
+        &self,
+        mut sectors_count: u64,
+        sector: u64,
+    ) -> result::Result<(), ExecuteError> {
+        sectors_count = sectors_count
+            .checked_add(sector)
+            .ok_or(ExecuteError::InvalidAccess)?;
+        if sectors_count > self.num_sectors() {
+            return Err(ExecuteError::InvalidAccess);
+        }
+        Ok(())
+    }
+
+    /// Executes `request` Request on `D` and `mem`.
+    pub fn execute<M: GuestMemory>(
```

Comment:
From a high level, it seems more convenient if the executor writes the status of the request as well. Does that make sense? I that case we probably no longer need `Unsupported` as an error variant, but haven't really looked into the details.